doge-evm/polygon-edge
# Remove offline validator from set in current epoch for simple punishment

# Description

When a new proposer is selected, other validators wait for its propose message. They will remove it from valid validator set if the propose message receiving timeout, set round change (+1) and move on. The proposer who removed from local cache would not be appear on list in this epoch.

In the last block of epoch, other validators would clear their offline validator set after a valid block committed, and move on to the next epoch.

# Changes include


# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist


## Testing


### Manual tests

1. Set up a 4-nodes local test network, which set `--epoch-size=30` for shorter confirmation. 
2. Create a new validator node, which staked for its valid validator authority.
3. Start the new validator node, watching the network's validator set and logs in 2 epoch.
4. Stop the new validator node, watching the network's validator set and logs in 2 epoch.
5. It should turn out that offline validator would not block the sealing work too much, and re-online validator would be able to get back to active in 1-2 epoch, which depends on its recover timestamp.
